documentation/examples/custom-sd/adapter-usage/main.go

```diff
@@ -0,0 +1,156 @@
+	outputFile    = a.Flag("output.file", "Output file for file_sd compatible file.").Default("/tmp/custom_sd.json").String()
+	listenAddress = a.Flag("listen.address", "The address the HTTP sd is listening on for requests.").Default("localhost:8080").String()
+	logFile       = a.Flag("log.file", "Location of logfile.").String()
+	logger        log.Logger
+)
+
+// Note: create a config struct for your custom SD type here.
+type sdConfig struct {
+	listenAddress   string
+	refreshInterval int
+	name            string
+}
+
+// Note: This is the struct with your implementation of the Discoverer interface (see Run function).
+type discovery struct {
+	listenAddress   string
+	refreshInterval time.Duration
+	name            string
+}
+
+// Note: The HTTP discovery in this example returns the targets it knows about as JSON.
+// ex: {"euw":["webserver"],"hq-test":["webserver","db"],"use":["webserver"]}
+// We need to convert this JSON to our internal type targetgroup.Group so that the
+// Adapter code, and eventually file_sd, knows how to deal with it.
+func (d *discovery) jsonToTargets(data []byte) ([]*targetgroup.Group, error) {
+	var arbJSON interface{}
+	var tgs []*targetgroup.Group
+
+	err := json.Unmarshal(data, &arbJSON)
+	if err != nil {
+		return nil, err
+	}
+
+	targetsMap := arbJSON.(map[string]interface{})
+	for k, v := range targetsMap {
+		targets := v.([]interface{})
+		targetsLabels := []model.LabelSet{}
+		tg := targetgroup.Group{Source: d.name}
+		for _, target := range targets {
+			targetLabel := model.LabelSet{
+				model.AddressLabel: model.LabelValue(target.(string)),
+			}
+			targetsLabels = append(targetsLabels[:], targetLabel)
+		}
+		labels := model.LabelSet{}
+		labels["dc"] = model.LabelValue(k)
+		tg.Labels = labels
+		tg.Targets = targetsLabels
+		// for our http sd we use the dc name as the source
+		tg.Source = k
+		tgs = append(tgs, &tg)
+	}
+	return tgs, nil
+}
+
+// Note: you must implement this function for your discovery implementation as part of the
+// Discoverer interface.
+func (d *discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+	targetsURL := fmt.Sprintf("http://%s/targets", d.listenAddress)
+
+	for c := time.Tick(d.refreshInterval); ; <-c {
+		resp, err := http.Get(targetsURL)
+		if err != nil {
+			continue
+		}
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			continue
+		}
+
+		tgs, err := d.jsonToTargets(body)
+		if err != nil {
+			logger.Log("error", err.Error())
+		} else {
+			ch <- tgs
+		}
+	}
+}
+
+func newDiscovery(cfg sdConfig) *discovery {
+	return &discovery{
+		listenAddress:   cfg.listenAddress,
+		refreshInterval: time.Second * time.Duration(cfg.refreshInterval),
+	}
+}
+
+func main() {
+	a.Parse(os.Args[1:])
+	a.HelpFlag.Short('h')
+
+	_, err := a.Parse(os.Args[1:])
+	if err != nil {
+		fmt.Println("err: ", err)
+		return
+	}
+
+	if *logFile != "" {
```

Comment:
This all should go. We don't offer logging to file, and it's not relevant to the example.